Add details for web API error handling (#14366)
aspnet/AspNetCore.Docs: aspnetcore/web-api/error-handling.md

[!code-csharp[](index/samples/2.x/Startup.cs?name=snippet_DisableProblemDetailsInvalidModelStateResponseFactory&highlight=5-14)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

## Client error response

For web API controllers, MVC transforms an error result (a result with status code 400 or higher) to a result with <xref:Microsoft.AspNetCore.Mvc.ProblemDetails>. The error response can be configured in one of the following ways:

1. [Use ApiBehaviorOptions.ClientErrorMapping](#use-apibehavioroptionsclienterrormapping)
1. [Implement ProblemDetailsFactory](#implement-problemdetailsfactory)

### Use ApiBehaviorOptions.ClientErrorMapping

Use the <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.ClientErrorMapping*> property to configure the contents of the `ProblemDetails` response. For example, the following code updates the `type` property for 404 responses:

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[](index/samples/3.x/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=8-9)]

::: moniker-end

::: moniker range="<= aspnetcore-2.2"

[!code-csharp[](index/samples/2.x/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=9-10)]

::: moniker-end

> ### Implement ProblemDetailsFactory

Reviewer: @pranavkm Is `ProblemDetailsFactory` available only in 3.0?

Author: Yup. It's new in 3.0

---
title: Error handling with web APIs
author: pranavkm

> description: Learn about error handling with web APIs.

Reviewer: I vote no. Web API has been there since the beginning of ASP.NET Core.

> description: Learn about error handling with web APIs.

Reviewer 2: Ok ... but the Dev Exception Page and bits like that ... seems like there's an awful lot of Core thingies here, it's a Core topic, and <2.1 is depreciated (at least) if not outright unsupported (i.e. 1.x bits).

> description: Learn about error handling with web APIs.

Reviewer: On 2nd thought, you're right. Not enough coffee yet. Hang tight while I wrap up a PR to patch this PR.

```
> curl https://localhost:5001/weatherforecast
System.ArgumentException: count
   at errorhandling.Controllers.WeatherForecastController.Get(Int32 x) in D:\work\Samples\samples\aspnetcore\mvc\errorhandling\Controllers\WeatherForecastController.cs:line 35
   at lambda_method(Closure , Object , Object[] )
   at Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(Object target, Object[] parameters)
...
```

> [!WARNING]
> Enable the Developer Exception Page **only when the app is running in the Development environment**. You don't want to share detailed exception information publicly when the app runs in production. For more information on configuring environments, see <xref:fundamentals/environments>.

## Exception handler

In non-development environments, the [Exception Handling Middleware](xref:fundamentals/error-handling) can be used to produce an error payload. To use the Exception Handling Middleware:

> 1. In `Startup.Configure`, invoke `UseExceptionHandler` to use the middleware:

Reviewer 2: ```suggestion
1. In `Startup.Configure`, invoke <xref:Microsoft.AspNetCore.Builder.ExceptionHandlerExtensions.UseExceptionHandler*> to use the middleware:
```

    ```csharp
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseDeveloperExceptionPage();
        }
        else
        {
            app.UseExceptionHandler("/error");
        }
    }
    ```

1. Configure a controller action to respond to the `/error` route:

    ```csharp
    [ApiController]
    public class ErrorController : ControllerBase
    {
        [Route("/error")]
        public IActionResult Error() => Problem();
    }
    ```

The preceding `Error` action sends an [RFC7807](https://tools.ietf.org/html/rfc7807)-compliant payload to the client.

> The Exception Handling Middleware can also provide more detailed content-negotiated output in the local development environment. Use the following steps to produce a consistent payload format across development and production environments:

Reviewer 2: ```suggestion
Exception Handling Middleware can also provide more detailed content-negotiated output in the local development environment. Use the following steps to produce a consistent payload format across development and production environments:
```

    ```csharp
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        if (env.IsDevelopment())
        {
            app.UseExceptionHandler("/error-local-development");
        }
        else
        {
            app.UseExceptionHandler("/error");
        }
    }
    ```

    In the preceding code, the middleware is registered with:

    * A route of `/error-local-development` in the Development environment.
    * A route of `/error` in environments that aren't Development.
    
1. Apply attribute routing to controller actions:

    ```csharp
    [ApiController]
    public class ErrorController : ControllerBase
    {
        [Route("/error-local-development")]
        public IActionResult ErrorLocalDevelopment(
            [FromServices] IWebHostEnvironment webHostEnvironment)
        {
            if (!webHostEnvironment.IsDevelopment())
            {
                throw new InvalidOperationException(
                    "This shouldn't be invoked in non-development environments.");
            }
    
            var context = HttpContext.Features.Get<IExceptionHandlerFeature>();

>             return Problem(

Reviewer 2: I thought you placed a line prior to returns (?) ...

```suggestion

            return Problem(
```

                detail: context.Error.StackTrace,
                title: context.Error.Message);
        }
    
        [Route("/error")]
        public IActionResult Error() => Problem();
    }
    ```

## Use exceptions to modify the response

> You may want to modify the contents of the response from outside of the controller. In ASP.NET 4.x Web API, one way to do this was using the `HttpResponseException` type. ASP.NET Core doesn't include an equivalent type. Support for `HttpResponseException` can be added with the following steps:

Reviewer 2: Repo rules push back usually on use of 2nd person (but MS style is ok with it and it almost always leads to passive voice, which I'm not a big fan of). Anyway... if you want to factor it out (and I add the API link here) ...

```suggestion
The contents of the response can be modified from outside of the controller. In ASP.NET 4.x Web API, one way to do this was using the <xref:System.Web.Http.HttpResponseException> type. ASP.NET Core doesn't include an equivalent type. Support for `HttpResponseException` can be added with the following steps:
```

❓ Is that supposed to be "Web API" or "web API"?

    {
        public int Status { get; set; } = 500;
    
        public object Value { get; set; }
    }
    ```

1. Create an action filter named `HttpResponseExceptionFilter`:

    ```csharp
    public class HttpResponseExceptionFilter : IActionFilter, IOrderedFilter
    {
        public int Order { get; set; } = int.MaxValue - 10;
    
        public void OnActionExecuting(ActionExecutingContext context) {}
    
        public void OnActionExecuted(ActionExecutedContext context)
        {
            if (context.Exception is HttpResponseException exception)
            {
                context.Result = new ObjectResult(exception.Value)
                {
                    Status = exception.Status,
                };
                context.ExceptionHandled = true;
            }
        }
    }
    ```

1. In `Startup.ConfigureServices`, add the action filter to the filters collection:

    ```csharp
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddControllers(options => 
            options.Filters.Add(new HttpResponseExceptionFilter()));
    }
    ```

## Validation failure error response

> For web API controllers, MVC responds with a `ValidationProblemDetails` response type when model validation fails. MVC uses the results of <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.InvalidModelStateResponseFactory> to construct

Reviewer 2: ```suggestion
For web API controllers, MVC responds with a <xref:Microsoft.AspNetCore.Mvc.ValidationProblemDetails> response type when model validation fails. MVC uses the results of <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.InvalidModelStateResponseFactory> to construct the error response for a validation failure. The following example uses the factory to change the default response type to <xref:Microsoft.AspNetCore.Mvc.SerializableError>:
```

> the error response for a validation failure. The following example uses the factory to change the default response type to `SerializableError`:

Reviewer 2: Remove this line ... there's a return stuck on the end of Line 163. I've already compensated for the text on Line 164 by adding it to my suggestion for Line 163.

```suggestion
```

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[](index/samples/3.x/Startup.cs?name=snippet_DisableProblemDetailsInvalidModelStateResponseFactory&highlight=4-13)]

::: moniker-end

::: moniker range="= aspnetcore-2.2"

[!code-csharp[](index/samples/2.x/Startup.cs?name=snippet_DisableProblemDetailsInvalidModelStateResponseFactory&highlight=5-14)]

> ::: moniker-end

Reviewer 2: Not going to show anything for 2.1? ... shouldn't the text and section heading here be versioned >=2.2 leading into these code snippets?

Reviewer: Nice catch! There should be something here for 2.1. Looking at it right now.

## Client error response

An *error result* is defined as a result with an HTTP status code of 400 or higher. For web API controllers, MVC transforms an error result to a result with <xref:Microsoft.AspNetCore.Mvc.ProblemDetails>.

::: moniker range=">= aspnetcore-3.0"

The error response can be configured in one of the following ways:

1. [Use ApiBehaviorOptions.ClientErrorMapping](#use-apibehavioroptionsclienterrormapping)
1. [Implement ProblemDetailsFactory](#implement-problemdetailsfactory)

::: moniker-end

::: moniker range="<= aspnetcore-2.2"

The error response can be configured as outlined in the [Use ApiBehaviorOptions.ClientErrorMapping](#use-apibehavioroptionsclienterrormapping) section.

::: moniker-end

> ### Use ApiBehaviorOptions.ClientErrorMapping

Reviewer 2: Consider moving this section heading to inside the moniker range that ends on Line 189. That way it only shows up for >=3.0.

Then, remove that paragraph on Line 193 and the moniker range (Lines 191 and 195). The sentence on Line 199 can pick up with the coverage. (Drop Lines 191-195.)

It will all compose a bit better imo. ........... or not if ur too busy to mess with it. We can fix it up on a later PR for this topic.

Use the <xref:Microsoft.AspNetCore.Mvc.ApiBehaviorOptions.ClientErrorMapping*> property to configure the contents of the `ProblemDetails` response. For example, the following code updates the `type` property for 404 responses:

::: moniker range=">= aspnetcore-3.0"

[!code-csharp[](index/samples/3.x/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=8-9)]

::: moniker-end

::: moniker range="<= aspnetcore-2.2"

[!code-csharp[](index/samples/2.x/Startup.cs?name=snippet_ConfigureApiBehaviorOptions&highlight=9-10)]

::: moniker-end

::: moniker range=">= aspnetcore-3.0"

### Implement ProblemDetailsFactory

> In ASP.NET Core 3.0 or later, MVC uses `Microsoft.AspNetCore.Mvc.ProblemDetailsFactory` to produce all instances of `ProblemDetails` and `ValidationProblemDetails`. This includes client error responses, validation failure error responses, and

Reviewer 2: ```suggestion
MVC uses `Microsoft.AspNetCore.Mvc.ProblemDetailsFactory` to produce all instances of <xref:Microsoft.AspNetCore.Mvc.ProblemDetails> and <xref:Microsoft.AspNetCore.Mvc.ValidationProblemDetails>. This includes client error responses, validation failure error responses, and the `Microsoft.AspNetCore.Mvc.ControllerBase.Problem` and <xref:Microsoft.AspNetCore.Mvc.ControllerBase.ValidationProblem> helper methods.
```

> the `Microsoft.AspNetCore.Mvc.ControllerBase.Problem` and <xref:Microsoft.AspNetCore.Mvc.ControllerBase.ValidationProblem> helper methods.

Reviewer 2: There's another return on Line 217 that I removed in my suggestion.

```suggestion
```
